tutorials/custom_model.ipynb

```diff
@@ -0,0 +1,1014 @@
+      "text/plain": [
+       "<Figure size 640x480 with 1 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "ax = plot_toy_data(x=train_X, y=Y)\n",
+    "ax.scatter(\n",
+    "    train_X[:, 0].detach().numpy().squeeze(),\n",
+    "    train_X[:, 1].detach().numpy().squeeze(),\n",
+    "    zs=prob_regression_model(train_X).mean.detach().squeeze().numpy(),\n",
+    ")\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally, since our custom model is based off `Model` and `Posterior`, we can use both analytic and MC based acquisition functions for optimization."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "(tensor([[0., 0.]]), tensor(-0.1540))"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from botorch.acquisition.analytic import LogExpectedImprovement\n",
+    "from botorch.optim.optimize import optimize_acqf\n",
+    "\n",
+    "optimize_acqf(\n",
+    "    LogExpectedImprovement(model=prob_regression_model, best_f=Y.max()),\n",
+    "    bounds=bounds,\n",
+    "    q=1,\n",
+    "    num_restarts=5,\n",
+    "    raw_samples=10,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Before using `qLogExpectedImprovement` we need to register an appropriate sampler for the `TorchPosterior`. We can use the following code to create a `MCSampler` for any subclass of `torch.distributions.Distribution`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from botorch.sampling.base import MCSampler\n",
+    "from botorch.sampling.get_sampler import GetSampler\n",
+    "from botorch.sampling.stochastic_samplers import StochasticSampler\n",
+    "\n",
+    "\n",
+    "@GetSampler.register(distributions.Distribution)\n",
+    "def _get_sampler_torch(\n",
+    "    posterior: TorchPosterior,\n",
+    "    sample_shape: torch.Size,\n",
+    "    *,\n",
+    "    seed: Optional[int] = None,\n",
+    ") -> MCSampler:\n",
+    "    return StochasticSampler(sample_shape=sample_shape, seed=seed)"
```

Comment:
I am curious to know what the use case is for `StochasticSampler` then? And yes, happy to switch in `ForkedRNGSampler`, it also gives the correct result

Reply:
You can use `StochasticSampler` with a stochastic optimizer. In BoTorch, we tend to use deterministic optimizers, which requires making the acquisition function evaluations deterministic (an approach called sample-average approximation). But you can also use a stochastic optimizer, like Adam, to optimize the acquisition function. This could be done by passing `gen_candidates=gen_candidates_torch` to `optimize_acqf`. In that case, it'd be perfectly fine to use `StochasticSampler`.